### PR TITLE
Feature/widespace adapter

### PR DESCRIPTION
**Pull request by jonasmattsson1**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```
- contact email of the adapter’s maintainer
- [ ] official adapter submission

### CODE REVIEW

**Comment by jonasmattsson1** [comment]
The contact email of the adapter’s maintainer:
jonas.mattsson@widespace.com

File: test/spec/adapters/widespace_spec.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    adapter.callBids(BID_REQUEST);
+
+    pbjs.widespaceHandleCB(response);
+
+    return bidManager.addBidResponse.firstCall.args[1];
+  }
+
+  beforeEach(() => {
+    pbjs._bidsRequested = [];
```
Comment (matthewlane) [inline test/spec/adapters/widespace_spec.js]: This causes several other unit tests to fail

File: adapters.json [hunk 2]
```diff
@@ -35,6 +35,7 @@
     "centro",
     "roxot",
     "vertoz",
+    'widespace',
```
Comment (matthewlane) [inline adapters.json]: JSON requires double quotes, please use them here

File: test/spec/adapters/widespace_spec.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  }];
+
+
+describe('WidespaceAdapter', () => {
+
+  let adapter;
+  let sandbox;
+
+  beforeEach(() => {
+    pbjs._bidsRequested = [];
```
Comment (matthewlane) [inline test/spec/adapters/widespace_spec.js]: This can cause other test to fail, please remove

File: test/spec/adapters/widespace_spec.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    beforeEach(() => {
+      sandbox.stub(bidManager, 'addBidResponse');
+      sandbox.stub(adLoader, 'loadScript');
+
+      adapter.callBids(BID_REQUEST);
+      pbjs._bidsRequested.push(BID_REQUEST);
+      pbjs.widespaceHandleCB(BID_RESPONSE);
+
+      console.log(bidManager.addBidResponse.firstCall.args[0]);
```
Comment (matthewlane) [inline test/spec/adapters/widespace_spec.js]: Please remove console.log statements from tests

File: test/spec/adapters/widespace_spec.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+      successfulBid1 = bidManager.addBidResponse.firstCall.args[1];
+    });
+
+    it('should add one bid', () => {
+      sinon.assert.calledOnce(bidManager.addBidResponse);
+    });
+
+    it('should use the CPM returned by the server', () => {
+      console.log('successfulBid1', successfulBid1);
```
Comment (matthewlane) [inline test/spec/adapters/widespace_spec.js]: Remove

File: test/spec/adapters/widespace_spec.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      console.log('successfulBid1 CPM :::', successfulBid1.cpm);
```
Comment (matthewlane) [inline test/spec/adapters/widespace_spec.js]: Remove

File: test/spec/adapters/widespace_spec.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  describe('widespaceHandleCB', () => {
+    it('should exist and be a function', () => {
+      expect(pbjs.widespaceHandleCB).to.exist.and.to.be.a('function');
+    });
+  });
+
+
+  describe('respond with a successful bid', () => {
+    let successfulBid1,
+        placementCode;
```
Comment (matthewlane) [inline test/spec/adapters/widespace_spec.js]: This variable is never used, please remove if you don't need it

**Comment by jonasmattsson1** [comment]
Thank you, I have updated the tests.

You should now receive test ad´s with 
`{
  bidder: 'widespace',
  params: {
    sid: 'f666bfaf-69cf-4ed9-9262-08247bb274e4',
    cur: 'EUR'
  }
}`

File: test/spec/adapters/widespace_spec.js (hunk 6 above)
Comment (jonasmattsson1) [inline test/spec/adapters/widespace_spec.js]: Removed, thank you

File: test/spec/adapters/widespace_spec.js (hunk 5 above)
Comment (jonasmattsson1) [inline test/spec/adapters/widespace_spec.js]: Removed, thank you

File: test/spec/adapters/widespace_spec.js (hunk 4 above)
Comment (jonasmattsson1) [inline test/spec/adapters/widespace_spec.js]: Removed, thank you

File: test/spec/adapters/widespace_spec.js (hunk 7 above)
Comment (jonasmattsson1) [inline test/spec/adapters/widespace_spec.js]: Removed, thank you

File: test/spec/adapters/widespace_spec.js (hunk 3 above)
Comment (jonasmattsson1) [inline test/spec/adapters/widespace_spec.js]: Fixed, thank you

**Comment by matthewlane** [comment]
Changes look good and tests pass. I'm still getting back `noad` status responses with those params though, any pointers? Once I can verify a valid status response we should be good to merge

**Comment by jonasmattsson1** [comment]
Hi,

I think you get  status = 'noad' because we don´t have 300x250 or 300x600 ad sizes in our system.
I added [300,300] to `adunit.sizes`. Hope that is okey.

Thank you,
Jonas

**Comment by jonasmattsson1** [comment]
Hi, Can you try to emulate a Phone like iPhone in Chrome for example.
It´s probably that I think, We only serve ad´s to mobile devices.

**Comment by matthewlane** [comment]
No luck with an iPhone or Chrome on mobile for me. Here's a test page I've set up http://client-test.devnxs.net/widespace/. This is using prebid.js built with your adapter. Currently I'm seeing
```
available:  []
empty or error:  ["widespace"]
```
in the dev console. For review I check if the adapter is showing up in the `available` category. Can you check this page as well? Let me know if you are getting a different result, if the params I'm using on the html in that page look wrong, etc. Hopefully this can help track down why I'm not seeing bids. Thanks!

**Comment by jonasmattsson1** [comment]
Thank you for the time to set up the test page. I got noad on your test page aswell but 
I see that you are using widespace param  sid = "f666bfaf-69cf-4ed9-9262-08247bb274e4", It should be sid = "7b6589bf-95c8-4656-90b9-af9737bb9ad3" that are linked to a demo account.

I get ad´s on your page when I rewrite the sid parameter to the right one in Charles, see the print screen  that I made.

[https://drive.google.com/file/d/0B2Hq0o69IghYRS0yTXU0eUJ4dEU/view?usp=sharing](url)
You still need to emulate an mobile device.

Thank you,
Jonas

**Comment by matthewlane** [comment]
Getting bids now with "7b6589bf-95c8-4656-90b9-af9737bb9ad3" and mobile emulation, thanks

**Pull request merged**